### PR TITLE
Refactor <EmailData> in Domains Management to remove sites-list

### PR DESCRIPTION
**Pull request by dmsnell**
Part of https://github.com/Automattic/wp-calypso/projects/3

This PR refactors the `<EmailData />` component in domain management under **Domains** > **G Suite**

It no longer relies on `sites-list` and it no longer incorporates an `observe()` mixing.

There should be no functional or visible changes with this PR.

**Testing**
Please navigate to the domain settings pages and smoke-test with sites which are and sites which aren't connected to _G Suite_. Switch sites and make sure things update properly. We should try and make sure that product updates from the API update this page but I'm not sure how to do that.

**Notes**
 - This component is still pretty messy. The store connection is an obstacle in cleaning it up but is a big project in its own right easily deferrable until later.
 - There's some funky business going on in `connect()` because the dispatch-able functions depend on app state. I wasn't sure how to clean that up since I don't understand all the inner-workings of the components and so I hooked it all up in `mergeProps`. There has got to be a cleaner and simpler way to hook up this component.

cc: @gwwar (@umurkontaci @drewblaisdell @gziolo since y'all were on the history in that component)

### CODE REVIEW

File: client/components/data/domain-management/email/index.jsx [hunk 1]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
...
+	const selectedSite = getSelectedSite( state );
+	const googleAppsUsers = selectedDomainName
+		? getByDomain( state, selectedDomainName )
+		: getBySite( state, selectedSite.ID );
+
+	return {
+		googleAppsUsers,
+		googleAppsUsersLoaded: isLoaded( state ),
+		products: getProductsList( state ),
+		selectedSite,
```
Comment (SpenUK) [inline client/components/data/domain-management/email/index.jsx]: A bit of a nit-pick here, but selectedSite[ should be up top](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#object-property-shorthand) with googleAppsUsers right?

Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: Yes. I had them the other way in order to be alphabetical, which I feel is better, but it does go against our style guide. Thanks @SpenUK 

updated in eece3d5

Comment (SpenUK) [inline client/components/data/domain-management/email/index.jsx]: Ahh! I knew there'd be some logical reason :D

File: client/components/data/domain-management/email/index.jsx [hunk 2]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
 	};
 }
 
-const EmailData = React.createClass( {
-	displayName: 'EmailData',
+export class EmailData extends Component {
+	static propTypes = {
```
Comment (gziolo) [inline client/components/data/domain-management/email/index.jsx]: `component: React.PropTypes.func.isRequired` should still be a part of `propTypes`.

File: client/components/data/domain-management/email/index.jsx [hunk 3]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
...
-		sitePlans: React.PropTypes.object.isRequired,
-		sites: React.PropTypes.object.isRequired,
-		googleAppsUsers: React.PropTypes.array.isRequired,
-		googleAppsUsersLoaded: React.PropTypes.bool.isRequired
-	},
+	constructor( props ) {
+		super( props );
 
-	mixins: [ observe( 'productsList' ) ],
+		props.loadDomainsAndSitePlans();
```
Comment (gziolo) [inline client/components/data/domain-management/email/index.jsx]: I see for the first time a network call in the ES6 class constructor call. I couldn't find anything about related common practices in React docs (https://facebook.github.io/react/docs/reusable-components.html#es6-classes). I'm just leaving a comment to make sure we double check if it's okey or rather we should use `componentDidMount` method to do it. It looks like constructor's role is to set initial state and explicitly bind handlers to `this`.

File: client/components/data/domain-management/email/index.jsx [hunk 4]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
...
+const mapStateToProps = ( state, { selectedDomainName } ) => {
+	const selectedSite = getSelectedSite( state );
+	const googleAppsUsers = selectedDomainName
+		? getByDomain( state, selectedDomainName )
+		: getBySite( state, selectedSite.ID );
+
+	return {
+		googleAppsUsers,
+		selectedSite,
+		googleAppsUsersLoaded: isLoaded( state ),
```
Comment (gziolo) [inline client/components/data/domain-management/email/index.jsx]: It's not related to this PR, but in this context `isLoaded` method name doesn't feel right :)

File: client/components/data/domain-management/email/index.jsx (hunk 3 above)
Comment (gwwar) [inline client/components/data/domain-management/email/index.jsx]: Let's slim this down to just `fetchDomains` since we have a query component for site plans. We can then remove some of the manual plan dispatches below.
```
import QuerySitePlans from 'components/data/query-site-plans';
<QuerySitePlans siteId={ siteId } />
```

File: client/components/data/domain-management/email/index.jsx [hunk 5]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
...
+	if ( ! shouldFetchSitePlans( sitePlans, selectedSite ) ) {
+		return;
 	}
+
+	fetchDomains( selectedSite.ID );
+	fetchSitePlans( selectedSite.ID );
+};
+
+const mapStateToProps = ( state, { selectedDomainName } ) => {
+	const selectedSite = getSelectedSite( state );
```
Comment (gwwar) [inline client/components/data/domain-management/email/index.jsx]: If we update the sitePlans selector to use `getPlansBySiteId`, we should be able to get away with just fetching the selected site id.
```
const selectedSiteId = getSelectedSiteId( state );
```

File: client/components/data/domain-management/email/index.jsx (hunk 4 above)
Comment (gwwar) [inline client/components/data/domain-management/email/index.jsx]: We could add an alias when it's imported, eg `import  { isLoaded as isGoogleAppsUsersLoaded } from 'state/google-apps-users/selectors'`

**Comment by gwwar** [comment]
This one is going to be pretty hard to untangle completely, but we can do this step by step. 

If you're comfortable with it could we maybe try untangling the site plan fetches in this PR, and I'll try to test this again when that's ready.

File: client/components/data/domain-management/email/index.jsx (hunk 3 above)
Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: This is kind of ugly because the original component was doing it @gziolo. I like what you suggested @gwwar and will try and pull in the site plans query component.

It's hard for me making the subjective call between _cleanup component_ and _eliminate sites-list_.

File: client/components/data/domain-management/email/index.jsx [hunk 6]
```diff
@@ -48,84 +51,125 @@ function getStateFromStores( props ) {
...
-				if ( shouldFetchSitePlans( sitePlans, site ) ) {
-					dispatch( fetchSitePlans( site.ID ) );
-				}
-			}
-		}
+	if ( ! shouldFetchSitePlans( sitePlans, selectedSite ) ) {
+		return;
 	}
+
+	fetchDomains( selectedSite.ID );
```
Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: we're still using `selectedSite` though in `mergeProps` because `fetchGoogleAppsUsers` is dependent on it

File: client/components/data/domain-management/email/index.jsx (hunk 5 above)
Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: we're still using `selectedSite` though in `mergeProps` because `fetchGoogleAppsUsers` is dependent on it

File: client/components/data/query-site-plans/index.jsx [hunk 7]
```diff
@@ -24,7 +24,7 @@ class QuerySitePlans extends Component {
 		}
 	}
 
-	componentWillMount() {
+	componentDidMount() {
```
Comment (gwwar) [inline client/components/data/query-site-plans/index.jsx]: Why did you need to change this?

File: client/components/data/domain-management/email/index.jsx [hunk 8]
```diff
@@ -22,110 +22,117 @@ import {
...
-
-			this.prevSelectedSite = selectedSite;
-		}
-	},
+	}
 
 	render() {
+		const {
+			selectedSite: { ID: siteId },
+		} = this.props;
```
Comment (gwwar) [inline client/components/data/domain-management/email/index.jsx]: We could just pass this through as a prop directly

File: client/components/data/domain-management/email/index.jsx [hunk 9]
```diff
@@ -22,110 +22,117 @@ import {
...
-				stores={ stores }
-				getStateFromStores={ getStateFromStores }
-				products={ this.props.productsList.get() }
-				selectedDomainName={ this.props.selectedDomainName }
-				selectedSite={ this.props.sites.getSelectedSite() }
-				sitePlans={ this.props.sitePlans }
-				context={ this.props.context } />
+			<div>
+				<QueryProducts />
+				<QuerySiteDomains { ...{ siteId } } />
```
Comment (gwwar) [inline client/components/data/domain-management/email/index.jsx]: A little overkill on the spread operator? 😄

Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: this is a pattern I have used for a long time to eliminate needless redundancies in the code. writing something like `siteId={ siteId }` bugs me, especially when it's like five or more values all of the same name as the attribute

File: client/components/data/store-connection/index.jsx [hunk 10]
```diff
@@ -58,17 +58,22 @@ const StoreConnection = React.createClass( {
 	},
 
 	render() {
+		const props = {
+			...this.props,
```
Comment (dmsnell) [inline client/components/data/store-connection/index.jsx]: This is something that could probably go into a different PR, but I noticed some strange behaviors in the store connector and I felt that this and a couple of other changes would clean it up quite a bit.

By including the props here it means we don't have to duplicate all of the props in `getPropsFromState` or whatever it is that this calls to generate props. 

After some investigation I realized that about half of this file could be zapped anyway.

File: client/state/products-list/selectors.js [hunk 11]
```diff
@@ -1,3 +1,7 @@
+import { property } from 'lodash';
+
+export const getProductsList = property( 'productsList.items' );
```
Comment (dmsnell) [inline client/state/products-list/selectors.js]: uh…what do you think we should test?

I think the test would end up being circular, as in we'll either mock out this very structure and test that it matches (which doesn't actually test anything but does add maintenance overhead) or try and instantiate the entire Redux tree inside the test (which is slow and prone to fail for a host of unrelated reasons)

if there's something I'm missing I'm open to hearing it, but I think trying to force a test for this is a net loss for us.

File: client/components/data/domain-management/email/index.jsx (hunk 8 above)
Comment (dmsnell) [inline client/components/data/domain-management/email/index.jsx]: I was working on seeing if we could take out `selectedSite`, but it's coming in as a prop anyway right now for other uses and I don't see a reason to duplicate here by adding `siteId` as an additional prop.

File: client/components/data/store-connection/index.jsx (hunk 10 above)
Comment (gwwar) [inline client/components/data/store-connection/index.jsx]: Up to you, though this does increase the chance of regression.

Comment (dmsnell) [inline client/components/data/store-connection/index.jsx]: > increase the chance of regression.

yeah I'll be reviewing that before pushing this out. like I said, I have this component targeted for another PR. thanks for the catch!

File: client/state/products-list/selectors.js (hunk 11 above)
Comment (dmsnell) [inline client/state/products-list/selectors.js]: > Why would we do this, when we can use test data?

well that was my point. what test data do I use? I end up creating data to match the implementation of the selector in order to test that the implementation of the selector matches the test data, no?

I'm pretty opposed to testing that the code and structure we write matches the code and structure we write. I'm much more excited about testing behaviors.

```js
// file.js
export const CONSTANT = 5;

// test.js
import { CONSTANT } from 'file'

assert( CONSTANT === 5 );
```

Comment (gwwar) [inline client/state/products-list/selectors.js]: > what test data do I use?

What you expect your sub-tree to look like. I also find it useful to read tests to see what the general structure of what something looks like without having to spin up a branch.

>  I'm much more excited about testing behaviors.

Selectors are simple functions. We're just testing that we get the desired output for a given input. Ie what happens when an item exists, when happens when it doesn't, are we accidentally mutating data as we do this? 

Are you asking why we have tests in general?

Comment (dmsnell) [inline client/state/products-list/selectors.js]: I'm absolutely not against testing. Here's what I'm trying to say though…

```js
import { getProductsList } from '../selectors';

const mockState = { productsList: { items: [] } };

expect( getProductsList( mockState ) ).to.eql( [] );
```

That's the test here, and I've seen many tests very much like this in Calypso. I don't want to add them because they don't actually test anything. They are literally just duplicating code and doubling our maintenance burden. How did we know what kind of data to mock out? By looking at the implementation of the function we're trying to test. It's a circular dependency and it's a net loss for maintainability.

Put in other words, what happens if we were to change the structure of the state tree? Our test wouldn't fail. The test is built around the structure of the existing code. To make the change successful we need to start by updating both the test and the logic together.

Granted, this is not to say that we can't theoretically test things here but I would argue that we're not setup for it and to set ourselves up to properly test would require far more effort than this PR warrants.

Comment (gwwar) [inline client/state/products-list/selectors.js]: Yes, your example is a perfectly reasonable test. 

> Put in other words, what happens if we were to change the structure of the state tree? 

Our reducer tests would catch that. 😄 

We could go on about using fixtures or factories, but I really don't think it's a big deal to copy test data. It makes reading tests much easier. 

I guess where do you draw the line @dmsnell ? Do you wish we had no tests for any of our redux functions? I personally find and fix bugs when writing tests. Refactoring becomes much simpler too, for example in #7702 I was able to swap in another json validator rather easily since we had good persistence test coverage.

**Comment by gwwar** [comment]
Thanks for tackling this @dmsnell. I'd be fine with shipping this once we verify that there are no regressions with the G Suite Section. 

I set up a test site but it looks like it needs another 24 hrs to resolve the g suite settings. Let me see if there's a site that already has this.

**Comment by gwwar** [comment]
Looks like we're missing the setup notice:

In branch:
screen shot 2016-10-18 at 3 56 30 pm

vs stage/master:
screen shot 2016-10-18 at 3 57 29 pm

**Comment by dmsnell** [comment]
Thanks @gwwar for all your help and testing and suggestions. The testing part with **G Suite** is what I'm most concerned about being able to properly test.

I'm pretty close to scrapping yet another PR and trying to hit smaller.

**Comment by gwwar** [comment]
@dmsnell did you still need this one, or were you going to close this out?

**Comment by dmsnell** [comment]
No problem closing. It was just tracking changes and I had abandoned it of a smaller one. Thanks @gwwar
